src/storage/page_buffer.c

```diff
@@ -8215,11 +8211,56 @@ pgbuf_request_data_page_from_page_server (const VPID * vpid, log_lsa target_repl
   vpid_utils::pack (pac, *vpid);
   cublog::lsa_utils::pack (pac, target_repl_lsa);
 
-  std::string message (buffer.get (), size);
-  ts_Gl->push_request (tran_to_page_request::SEND_DATA_PAGE_FETCH, std::move (message));
+  std::string request_message (buffer.get (), size);
+
   if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
     {
-      _er_log_debug (ARG_FILE_LINE, "[READ DATA] Sent request for Page to Page Server. VPID: %d|%d\n", VPID_AS_ARGS (vpid));
+      _er_log_debug (ARG_FILE_LINE, "[READ DATA] Send request for Page to Page Server. VPID: %d|%d\n", VPID_AS_ARGS (vpid));
+    }
+
+  std::string response_message;
+  ts_Gl->send_receive (tran_to_page_request::SEND_DATA_PAGE_FETCH, std::move (request_message), response_message);
+
+  const char *message_buf = response_message.c_str ();
+  int error_code = NO_ERROR;
+  std::memcpy (&error_code, message_buf, sizeof (error_code));
+
+
+  // The message is either an error code or the content of the data page
+  if (response_message.size () == sizeof (int))
+    {
+      // We have an error.
+      int error_code;
+      std::memcpy (&error_code, response_message.c_str (), sizeof (error_code));
+
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
+	{
+	  _er_log_debug (ARG_FILE_LINE, "[READ DATA] Received error: %d \n", error_code);
+	}
+
+      if (error_code == ER_PB_BAD_PAGEID)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_PB_BAD_PAGEID, 2, VPID_AS_ARGS (vpid));
+	}
+      else
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 0);
+	}
+
+      return error_code;
```

Comment:
wipe `io_page` before return?

Reply:
I don't think it is necessary; caller should ignore io_page in case of errors.